File: solutions/ocaml/048-unique-paths.ml
(*
  Problem 48: Unique Paths (LeetCode 62)
  Difficulty: Med
  Language: OCaml
*)
let unique_paths m n =
  let k = min (m - 1) (n - 1) in
  let rec go i acc =
    if i >= k then acc
    else go (i + 1) (acc * (m + n - 2 - i) / (i + 1))
  in go 0 1

File: solutions/rust/048-unique-paths.rs
/**
 * Problem 48: Unique Paths (LeetCode 62)
 * Difficulty: Med
 * Language: Rust
 */
pub fn unique_paths(m: i32, n: i32) -> i32 {
    let k = m.min(n) - 1;
    (0..k).fold(1i64, |acc, i| acc * (m as i64 + n as i64 - 2 - i as i64) / (i as i64 + 1)) as i32
}
